### ISSUE TITLE
Reading a vector field as boost::multi_array<Eigen::Vector3f, 3>

### ISSUE THREAD

**Issue by acdemiralp** [BlueBrain/HighFive#256]
Dear all,

I wonder if this is possible, or any tips on how to make it possible. Currently I have to copy a boost::multi_array<float, 4> to a boost::multi_array<Eigen::Vector3f, 3> which kills performance on even GB data.

**Comment by alkino** [comment]
Hi,
Can you test with this patch: https://github.com/BlueBrain/HighFive/pull/258 ?
Regards

**Comment by alkino** [comment]
I tried with this little stupid example:
```
#include <iostream>
#include <highfive/H5Easy.hpp>
#include <Eigen/Eigen>

int main()
{
    {
        Eigen::Vector3f vec(1,2,3);
        H5Easy::File file("example.h5", H5Easy::File::Overwrite);

        H5Easy::dump(file, "/path/to/A", vec);

        auto a = H5Easy::load<Eigen::Vector3f>(file, "/path/to/A");
        std::cout << vec << std::endl;
    }
    return 0;
}
```

**Comment by acdemiralp** [comment]
The following static assert throws when I try to load the vector field on the #258 branch:
```cpp

// apply conversion operations to basic scalar type
template <typename Scalar, class Enable = void>
struct data_converter {
    inline data_converter(Scalar&, DataSpace&) {

        static_assert((std::is_arithmetic<Scalar>::value ||
                       std::is_enum<Scalar>::value ||
                       std::is_same<std::string, Scalar>::value),
                      "supported datatype should be an arithmetic value, a "
                      "std::string or a container/array");
    }
```

**Comment by acdemiralp** [comment]
Patching the following into H5Converter_misc.hpp at least compiles:
```cpp
#ifdef H5_USE_EIGEN
#include <Eigen/Eigen>
#endif
```
and
```cpp
#ifdef H5_USE_EIGEN
template <typename T, std::int32_t M, std::int32_t N>
struct data_converter<Eigen::Matrix<T, M, N>, void> {

    inline data_converter(Eigen::Matrix<T, M, N>&, DataSpace& space)
        : _dims(space.getDimensions()) {
        assert(_dims.size() == M * N);
    }

    inline typename type_of_array<T>::type* transform_read(Eigen::Matrix<T, M, N>& array) {
        return array.data();
    }

    inline typename type_of_array<T>::type* transform_write(Eigen::Matrix<T, M, N>& array) {
        return array.data();
    }

    inline void process_result(Eigen::Matrix<T, M, N>&) {}

    std::vector<size_t> _dims;
};
#endif
```
This does not work for dynamically sized matrices (when M and N are -1), and will massively frustrate the users which would like to use them, so please do not add to the main codebase.

**Comment by acdemiralp** [comment]
In order to get it to work with nested Eigen::Vector3fs in a boost::multi_array, I had to do the following (constexpr if statements require C++17):
```cpp
#ifdef H5_USE_BOOST
// apply conversion to boost multi array
template <typename T, std::size_t Dims>
struct data_converter<boost::multi_array<T, Dims>, void> {

    typedef typename boost::multi_array<T, Dims> MultiArray;

    inline data_converter(MultiArray&, DataSpace& space)
        : _dims(space.getDimensions()) {
        assert(_dims.size() == Dims);
    }

    inline typename type_of_array<T>::type* transform_read(MultiArray& array) {
        if (std::equal(_dims.begin(), _dims.end(), array.shape()) == false) {
            boost::array<typename MultiArray::index, Dims> ext;
            std::copy(_dims.begin(), _dims.end(), ext.begin());
            array.resize(ext);
        }
        if constexpr (std::is_arithmetic<T>::value || std::is_enum<T>::value || std::is_same<std::string, T>::value)
          return array.data();
        else
          return array.data()->data();
    }

    inline typename type_of_array<T>::type* transform_write(MultiArray& array) {
      if constexpr (std::is_arithmetic<T>::value || std::is_enum<T>::value || std::is_same<std::string, T>::value)
        return array.data();
      else
        return array.data()->data();
    }

    inline void process_result(MultiArray&) {}

    std::vector<size_t> _dims;
};
```

**Comment by alkino** [comment]
Yes, I discover that `Eigen` is handled only by the `Easy` part.
I have written tests with more coverage and I will try to make it work.
We have to use `.cols` and `.rows` instead of `M` and `N` for dynamic matrix.

**Comment by alkino** [comment]
@tdegeus I don't get why you implement support of Eigen only in Easy?
Why such a distinction?

**Comment by tdegeus** [comment]
@alkino It would be great if you'd be willing to implement Eigen (and possibly xtensor) deeper down in the code !

**Comment by tdegeus** [comment]
Though in the end I do think that the H5Easy API (`dump` and `load`) should stay.

**Comment by alexsavulescu** [comment]
@acdemiralp could you try with #261 and provide some feedback?

**Comment by alkino** [comment]
Should be fixed by #586
